packages/flutter/lib/src/widgets/async.dart

```diff
@@ -0,0 +1,256 @@
+  active,
+
+  /// Connected to a terminated asynchronous computation.
+  done
+}
+
+/// Strategy for building a Widget based on asynchronous interaction.
+typedef Widget AsyncWidgetBuilder<T>(
+    BuildContext context, AsyncSnapshot<T> snapshot);
+
+/// Widget that builds itself based on interaction with the specified [future].
+///
+/// The building strategy is defined by the given [builder].
+///
+/// Behaves identically to [StreamBuilder] configured with [future].asStream()
+/// when [future] is non-null.
+class FutureBuilder<T> extends StatefulWidget {
+  final Future<T> future;
+  final AsyncWidgetBuilder<T> builder;
+
+  FutureBuilder({Key key, this.future, this.builder}) : super(key: key);
+
+  @override
+  State<FutureBuilder<T>> createState() => new _FutureBuilderState<T>();
+}
+
+class _FutureBuilderState<T> extends State<FutureBuilder<T>> {
+  Stream<T> _stream;
+
+  @override
+  void initState() {
+    super.initState();
+    _subscribe();
+  }
+
+  @override
+  void didUpdateConfig(FutureBuilder<T> oldConfig) {
+    if (!identical(oldConfig.future, config.future)) {
+      _unsubscribe();
+      _subscribe();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) =>
+      new StreamBuilder<T>(stream: _stream, builder: config.builder);
+
+  @override
+  void dispose() {
+    _unsubscribe();
+    super.dispose();
+  }
+
+  void _subscribe() {
+    _stream = config.future?.asStream();
+  }
+
+  void _unsubscribe() {
+    _stream = null;
+  }
+}
+
+/// Widget that builds itself based on the latest snapshot of interaction with
+/// the specified [stream].
+///
+/// The building strategy is defined by the given [builder].
+class StreamBuilder<T> extends StreamFold<T, AsyncSnapshot<T>> {
+  final AsyncWidgetBuilder<T> builder;
+
+  StreamBuilder({Key key, Stream<T> stream, this.builder})
+      : super(key: key, stream: stream);
+
+  @override
+  AsyncSnapshot<T> initial() =>
+      new AsyncSnapshot<T>(ConnectionState.none, null, null);
+
+  @override
+  AsyncSnapshot<T> onConnecting(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(
+          ConnectionState.waiting, current.data, current.error);
+
+  @override
+  AsyncSnapshot<T> onData(AsyncSnapshot<T> current, T data) =>
+      new AsyncSnapshot<T>(ConnectionState.active, data, null);
+
+  @override
+  AsyncSnapshot<T> onError(AsyncSnapshot<T> current, dynamic error) =>
+      new AsyncSnapshot<T>(ConnectionState.active, null, error);
+
+  @override
+  AsyncSnapshot<T> onDone(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(ConnectionState.done, current.data, current.error);
+
+  @override
+  AsyncSnapshot<T> onDisconnecting(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(
+          ConnectionState.none, current.data, current.error);
+
+  @override
+  Widget build(BuildContext context, AsyncSnapshot<T> currentSummary) =>
+      builder(context, currentSummary);
+}
+
+/// Base class for Widgets that build themselves based on interaction with
```

Comment:
nit: please order the classes so that superclasses come first, so that someone reading the file reads the classes in the order in which members are introduced